### PR TITLE
Homework №3: Time

### PR DESCRIPTION
**Pull request by ImmortalTurtle**
Implemented `sleep()` and `uptime()`.
There is only one timer yet though. Planning to make a list of them.

I'm not sure about the interface of `uptime`: right now it takes buffer as first argument. Buffer is needed to store the string representation of time. Otherwise (if we define buffer in `sys_uptime()`) at the time of executing `os_sys_write()` program catches _Segmentation fault_.

That happens if we put `os_sys_write()` in any `sys_*` function. That doesn't seem right to me, but may be that is how it's supposed to be.
@AntonKozlov Do you consider this a problem? And what would you advice considering everything I've mentioned?

### CODE REVIEW

File: src/os/syscall.c [hunk 1]
```diff
@@ -183,6 +188,40 @@ static long sys_sem_free(int syscall,
...
+		sched_wait();
+		sched();
+	}
+
+	irq_enable(irq);
+
+	return 0;
+}
+
+static long sys_uptime(int syscall,
```
Comment (AntonKozlov) [inline src/os/syscall.c]: Minor: `char *` is not very convenient way to report uptime. You could decide to show it in human-readable in future; then you'll need to parse string first
I'd suppose to return `long` with microseconds:  `2 ** 63 / (1000_000 * 60 * 60 * 24 * 365) == >292471 years`; should be enough

File: src/os/time.c [hunk 2]
```diff
@@ -5,30 +5,70 @@
...
+struct timer *single;
+double init_time;
+
+double get_init_time() {
+	return init_time;
+}
+
+double get_current_time() {
+	struct timeval cur_time;
+	gettimeofday(&cur_time, NULL);
```
Comment (AntonKozlov) [inline src/os/time.c]: You're supposed to implement `gettimeofday`, suppose host `gettimeofday` is not available. You could use https://antonkozlov.github.io/cs226/7.html#9 as general scheme

File: src/os/time.c [hunk 3]
```diff
@@ -5,30 +5,70 @@
...
 	struct itimerval cur_it;
 	getitimer(ITIMER_REAL, &cur_it);
-	/*fprintf(stderr, "%s: tv_usec=%ld\n", __func__, cur_it.it_value.tv_usec);*/
+	//TODO: remove timer
+
+	//TODO: move into function
+	single->sec_left = 0;
+	if(single->task != NULL && single->task->state == SCHED_SLEEP) {
+		sched_notify(single->task);
+		sched();
```
Comment (AntonKozlov) [inline src/os/time.c]: it's seems bad place to call `sched`: we aren't exit from signal handler (violating signal handling procedure)

In general, you should modify user context to return to `sched_tramp` as made in os_sigiohnd. By implementation, you could refactor this part in common place

File: src/os/syscall.c (hunk 1 above)
Comment (AntonKozlov) [reply src/os/syscall.c]: Full signature would be something like `long os_uptime(void)`

**Comment by AntonKozlov** [comment]
> Otherwise (if we define buffer in sys_uptime()) at the time of executing os_sys_write() program catches Segmentation fault. That happens if we put os_sys_write() in any sys_* function. That doesn't seem right to me, but may be that is how it's supposed to be.

`os_sys_write` generate syscall to enter in kernel mode, but since we are in kernel mode already, the new syscall is probably is not handled. Even in case it worked, it's not correct from consistency and performance point of view. You should call `sys_write` for now or, even better, extract function to print in kernel mode

File: src/os/time.c (hunk 3 above)
Comment (ImmortalTurtle) [reply src/os/time.c]: Yeah, that's absolutely right, will fix.
Thank you

File: src/os/syscall.c (hunk 1 above)
Comment (ImmortalTurtle) [reply src/os/syscall.c]: Sure, now that I think of that it seems much more appropriate for `os_uptime()` to return time as long, not to print it.
And shell function `uptime` does the printing.

**Comment by ImmortalTurtle** [comment]
All right, seems legit to me. _Top half_ only sets the timer and *bot half* decrements and removes timers, notifies and calls `sched`. 
It's possible however to put a few more operations in _top half_ to simplify other code.

`bot_half_tramp` basically is copy of `sched_tramp` but firstly it calls `bot_half` instead of `sched`.

It's too bad that all this stuff with many timers and optimizations can't be tested right now. So I can only imagine it works exactly the way it did in my mind. :pensive:

File: src/os/time.c [hunk 4]
```diff
@@ -3,17 +3,143 @@
...
+	TAILQ_FOREACH(cur, &timers.head, link) {
+		if(tmr->usec_left < cur->usec_left) {
+			TAILQ_INSERT_BEFORE(cur, tmr, link);
+			return;
+		}
+	}
+	TAILQ_INSERT_TAIL(&timers.head, tmr, link);
+}
+
+int *new_timer(int seconds, struct sched_task *task, struct timer *tmr) {
```
Comment (AntonKozlov) [inline src/os/time.c]: Return value should be probably just `int`

File: src/os/time.c [hunk 5]
```diff
@@ -3,17 +3,143 @@
...
+#define INTERVAL_SEC 1
+#define INTERVAL_USEC 0
+
+extern void bot_half_tramp(void);
+
+struct {
+	TAILQ_HEAD(timerlist, timer) head;
+} timers;
+
+long uptime;
```
Comment (AntonKozlov) [inline src/os/time.c]: Private variables and functions should be `static` to make sure they aren't used

File: src/os/time.c [hunk 6]
```diff
@@ -3,17 +3,143 @@
...
+	sched();
+}
+
+static void os_sigalrmhnd(int signal, siginfo_t *info, void *ctx) {
+	top_half();
+
+	ucontext_t *uc = (ucontext_t *) ctx;
+	greg_t *regs = uc->uc_mcontext.gregs;
+
+	regs[REG_RSP] -= 8;
```
Comment (AntonKozlov) [inline src/os/time.c]: This bottom half is probably worth refactor to share as much code as possible with `sched_tramp`. But it's a flaw from my initial implementation

File: src/apps.c [hunk 7]
```diff
@@ -18,18 +21,18 @@ static int echo(int argc, char *argv[]) {
...
+	os_sleep(num);
 	return 0;
 }
 
 static int uptime(int argc, char *argv[]) {
-	/* FIXME print time passed from eduos kernel start, implement solely via eduos calls */
-	system("date +%s.%N");
+	char time_string[20];
+	double uptime = (double) os_uptime() / 1000000;
+	snprintf(time_string, ARRAY_SIZE(time_string), "%f", uptime);
```
Comment (AntonKozlov) [inline src/apps.c]: You could use `"%f\n"` format instead of `"%f"` and subsequent `strcat`

**Comment by AntonKozlov** [comment]
Oh, and forget to mention the sequence of commands like this make the shell (and may be the system) hang
```
> sleep 0
> sleep 1
```

`sleep 0` should be a reason, as subsequent sleeps are handled well. `sleep 0` itself seems be handled well too. 

Anyway, not a blocker

**Pull request merged**